File: core/integration/golden_vector.py
"""
BIZRA Golden Vector — Cross-Language Sealing Test (Python side)

This module defines the canonical test vector and verifies that Python
produces the identical hash as Rust for the same input.

The golden vector is FROZEN. If this hash changes, cross-language sealing
is broken and ALL receipts become untrustworthy.

Standing on: Merkle (1979), Aumasson (2015), Shannon (1948)

Usage:
    python -m pytest core/integration/golden_vector.py -v
    python core/integration/golden_vector.py  # Standalone verification
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

import blake3

# ── Constants (FROZEN — must match Rust bizra-core/src/canonical.rs) ──
DOMAIN_GOLDEN_VECTOR = "bizra-golden-vector-v1"
FIXED_POINT_P = 1_000_000  # 6 decimal places for float→int conversion

# ── The Golden Vector (FROZEN — never modify) ──────────────────────
# This is the canonical input that both Rust and Python must hash identically.
GOLDEN_MISSION_ID = "golden-vector-v1"
GOLDEN_INITIATOR_ID = "node0-genesis"
GOLDEN_PAYLOAD = b"In the Name of Allah, Most Gracious, Most Merciful"
GOLDEN_IHSAN_SCORE = 0.984700  # Fixed-point: round(0.984700 * 1_000_000) = 984700
GOLDEN_TIMESTAMP = 1711584000000  # 2024-03-28T00:00:00Z (fixed epoch)

# ── The Frozen Digest (computed once, verified forever) ──
# If this changes, cross-language sealing is broken.
# Frozen on 2026-03-28 after cross-language verification.
GOLDEN_DIGEST_HEX = "966725c27200cdd28632e1f10a09ca7f982491be5842c8eb1264650a32a51205"


@dataclass(frozen=True)
class GoldenVector:
    """The canonical cross-language test vector."""

    mission_id: str = GOLDEN_MISSION_ID
    initiator_id: str = GOLDEN_INITIATOR_ID
    payload: bytes = GOLDEN_PAYLOAD
    ihsan_score: float = GOLDEN_IHSAN_SCORE
    timestamp: int = GOLDEN_TIMESTAMP

    def serialize_canonical(self) -> bytes:
        """
        Deterministic serialization matching Rust's canonical_hasher.

        Field order: mission_id, initiator_id, payload, ihsan_fixed, timestamp
        Encoding:
          - Strings: UTF-8 bytes prefixed with u32le length
          - Bytes: raw bytes prefixed with u32le length
          - Float: round(value * FIXED_POINT_P) as u64le
          - Int: u64le
        """
        buf = bytearray()

        # mission_id: length-prefixed UTF-8
        mid = self.mission_id.encode("utf-8")
        buf.extend(struct.pack("<I", len(mid)))
        buf.extend(mid)

        # initiator_id: length-prefixed UTF-8
        iid = self.initiator_id.encode("utf-8")
        buf.extend(struct.pack("<I", len(iid)))
        buf.extend(iid)

        # payload: length-prefixed bytes
        buf.extend(struct.pack("<I", len(self.payload)))
        buf.extend(self.payload)

        # ihsan_score: fixed-point u64le (round, not truncate)
        ihsan_fixed = round(self.ihsan_score * FIXED_POINT_P)
        buf.extend(struct.pack("<Q", ihsan_fixed))

        # timestamp: u64le
        buf.extend(struct.pack("<Q", self.timestamp))

        return bytes(buf)

    def compute_hash(self) -> str:
        """
        Domain-separated BLAKE3 hash matching Rust canonical.rs::domain_hash.

        hash = BLAKE3(domain + ":" + serialized_data)
        """
        serialized = self.serialize_canonical()
        hasher = blake3.blake3()
        hasher.update(DOMAIN_GOLDEN_VECTOR.encode("utf-8"))
        hasher.update(b":")
        hasher.update(serialized)
        return hasher.hexdigest()


def compute_golden_digest() -> str:
    """Compute the golden vector digest."""
    vector = GoldenVector()
    return vector.compute_hash()


def verify_golden_digest(expected_hex: str | None = None) -> bool:
    """Verify the golden vector produces the expected digest."""
    actual = compute_golden_digest()
    if expected_hex is None:
        print(f"GOLDEN_DIGEST={actual}")
        return True
    return actual == expected_hex


# ═══════════════════════════════════════════════════════════════════
# TESTS (run with pytest)
# ═══════════════════════════════════════════════════════════════════


def test_golden_vector_produces_digest():
    """The golden vector must produce a non-empty, deterministic digest."""
    digest = compute_golden_digest()
    assert len(digest) == 64  # BLAKE3 hex = 64 chars
    # Run twice — must be identical (I₁ idempotency)
    assert digest == compute_golden_digest()


def test_golden_vector_serialization_deterministic():
    """Serialization must be byte-identical across invocations."""
    v1 = GoldenVector().serialize_canonical()
    v2 = GoldenVector().serialize_canonical()
    assert v1 == v2


def test_golden_vector_fixed_point_rounding():
    """Fixed-point conversion must round, not truncate."""
    # 0.984700 * 1_000_000 = 984700.0 — exact
    ihsan_fixed = round(GOLDEN_IHSAN_SCORE * FIXED_POINT_P)
    assert ihsan_fixed == 984700


def test_golden_vector_rejects_modified_input():
    """Changing any field must change the digest."""
    original = compute_golden_digest()

    # Tamper with payload
    tampered = GoldenVector(payload=b"Modified payload")
    assert tampered.compute_hash() != original

    # Tamper with ihsan
    tampered = GoldenVector(ihsan_score=0.50)
    assert tampered.compute_hash() != original

    # Tamper with mission_id
    tampered = GoldenVector(mission_id="tampered-id")
    assert tampered.compute_hash() != original


def test_golden_vector_matches_frozen_digest():
    """The digest must match the frozen canonical value."""
    digest = compute_golden_digest()
    assert digest == GOLDEN_DIGEST_HEX, (
        f"FATAL: Golden vector digest changed!\n"
        f"Expected: {GOLDEN_DIGEST_HEX}\n"
        f"Got:      {digest}\n"
        f"Cross-language sealing is BROKEN."
    )


def test_golden_vector_domain_separation():
    """Different domains must produce different hashes for same data."""
    vector = GoldenVector()
    serialized = vector.serialize_canonical()

    # Standard domain
    h1 = blake3.blake3(DOMAIN_GOLDEN_VECTOR.encode() + b":" + serialized).hexdigest()

    # Different domain
    h2 = blake3.blake3(b"different-domain:" + serialized).hexdigest()

    assert h1 != h2


if __name__ == "__main__":
    digest = compute_golden_digest()
    print(f"Python Golden Vector Digest: {digest}")
    print("HASH_ALGO: blake3")
    print(f"DOMAIN: {DOMAIN_GOLDEN_VECTOR}")
    print(f"Serialized length: {len(GoldenVector().serialize_canonical())} bytes")
    print()

    # Run self-tests
    test_golden_vector_produces_digest()
    test_golden_vector_serialization_deterministic()
    test_golden_vector_fixed_point_rounding()
    test_golden_vector_rejects_modified_input()
    test_golden_vector_domain_separation()
    print("All golden vector tests PASSED")

File: bizra-omega/bizra-core/src/golden_vector.rs
//! # BIZRA Golden Vector — Cross-Language Sealing Test (Rust side)
//!
//! This module defines the canonical test vector and verifies that Rust
//! produces the identical hash as Python for the same input.
//!
//! The golden vector is FROZEN. If this hash changes, cross-language sealing
//! is broken and ALL receipts become untrustworthy.
//!
//! Standing on: Merkle (1979), Aumasson (2015), Shannon (1948)

use blake3::Hasher;

/// Domain prefix for golden vector hashing.
pub const DOMAIN_GOLDEN_VECTOR: &str = "bizra-golden-vector-v1";

/// Fixed-point precision multiplier (P = 1,000,000).
pub const FIXED_POINT_P: f64 = 1_000_000.0;

// ── The Golden Vector (FROZEN — never modify) ──────────────
/// Frozen mission ID for cross-language sealing test.
pub const GOLDEN_MISSION_ID: &str = "golden-vector-v1";
/// Frozen initiator ID (node0 genesis identity).
pub const GOLDEN_INITIATOR_ID: &str = "node0-genesis";
/// Frozen payload (Basmala — first words of every Surah).
pub const GOLDEN_PAYLOAD: &[u8] = b"In the Name of Allah, Most Gracious, Most Merciful";
/// Frozen Ihsan score for deterministic hashing.
pub const GOLDEN_IHSAN_SCORE: f64 = 0.984700;
/// Frozen timestamp (2024-03-28T00:00:00Z) for reproducibility.
pub const GOLDEN_TIMESTAMP: u64 = 1711584000000;

/// The FROZEN digest. Python and Rust must produce this exact hash.
pub const GOLDEN_DIGEST_HEX: &str =
    "966725c27200cdd28632e1f10a09ca7f982491be5842c8eb1264650a32a51205";

/// Serialize the golden vector into canonical bytes.
///
/// Field order: mission_id, initiator_id, payload, ihsan_fixed, timestamp
/// Encoding:
///   - Strings: UTF-8 bytes prefixed with u32le length
///   - Bytes: raw bytes prefixed with u32le length
///   - Float: round(value * FIXED_POINT_P) as u64le
///   - Int: u64le
pub fn serialize_golden_vector() -> Vec<u8> {
    let mut buf = Vec::with_capacity(128);

    // mission_id: length-prefixed UTF-8
    let mid = GOLDEN_MISSION_ID.as_bytes();
    buf.extend_from_slice(&(mid.len() as u32).to_le_bytes());
    buf.extend_from_slice(mid);

    // initiator_id: length-prefixed UTF-8
    let iid = GOLDEN_INITIATOR_ID.as_bytes();
    buf.extend_from_slice(&(iid.len() as u32).to_le_bytes());
    buf.extend_from_slice(iid);

    // payload: length-prefixed bytes
    buf.extend_from_slice(&(GOLDEN_PAYLOAD.len() as u32).to_le_bytes());
    buf.extend_from_slice(GOLDEN_PAYLOAD);

    // ihsan_score: fixed-point u64le (round, not truncate)
    let ihsan_fixed = (GOLDEN_IHSAN_SCORE * FIXED_POINT_P).round() as u64;
    buf.extend_from_slice(&ihsan_fixed.to_le_bytes());

    // timestamp: u64le
    buf.extend_from_slice(&GOLDEN_TIMESTAMP.to_le_bytes());

    buf
}

/// Compute the golden vector hash using domain-separated BLAKE3.
///
/// `hash = BLAKE3(domain + ":" + serialized_data)`
pub fn compute_golden_digest() -> [u8; 32] {
    let serialized = serialize_golden_vector();
    let mut hasher = Hasher::new();
    hasher.update(DOMAIN_GOLDEN_VECTOR.as_bytes());
    hasher.update(b":");
    hasher.update(&serialized);
    hasher.finalize().into()
}

/// Compute and return the golden vector hex digest.
pub fn golden_digest_hex() -> String {
    compute_golden_digest()
        .iter()
        .map(|b| format!("{b:02x}"))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_golden_vector_produces_frozen_digest() {
        let digest = golden_digest_hex();
        assert_eq!(
            digest, GOLDEN_DIGEST_HEX,
            "FATAL: Golden vector digest mismatch. Cross-language sealing broken.\n\
             Expected: {}\n\
             Got:      {}",
            GOLDEN_DIGEST_HEX, digest
        );
    }

    #[test]
    fn test_golden_vector_serialization_deterministic() {
        let a = serialize_golden_vector();
        let b = serialize_golden_vector();
        assert_eq!(a, b, "Serialization is not deterministic");
    }

    #[test]
    fn test_golden_vector_serialization_length() {
        let serialized = serialize_golden_vector();
        // 4 + 16 + 4 + 13 + 4 + 50 + 8 + 8 = 107 bytes
        assert_eq!(serialized.len(), 107, "Serialized length mismatch");
    }

    #[test]
    fn test_golden_vector_fixed_point_rounding() {
        let ihsan_fixed = (GOLDEN_IHSAN_SCORE * FIXED_POINT_P).round() as u64;
        assert_eq!(ihsan_fixed, 984700, "Fixed-point rounding incorrect");
    }

    #[test]
    fn test_golden_vector_idempotency() {
        let d1 = golden_digest_hex();
        let d2 = golden_digest_hex();
        assert_eq!(d1, d2, "Hash is not idempotent");
    }

    #[test]
    fn test_golden_vector_domain_separation() {
        let serialized = serialize_golden_vector();

        // Standard domain
        let mut h1 = Hasher::new();
        h1.update(DOMAIN_GOLDEN_VECTOR.as_bytes());
        h1.update(b":");
        h1.update(&serialized);
        let d1: [u8; 32] = h1.finalize().into();

        // Different domain
        let mut h2 = Hasher::new();
        h2.update(b"different-domain");
        h2.update(b":");
        h2.update(&serialized);
        let d2: [u8; 32] = h2.finalize().into();

        assert_ne!(d1, d2, "Domain separation failed");
    }
}
